osquery/tables/system/posix/augeas.cpp

```diff
@@ -130,50 +81,71 @@ void matchAugeasPattern(augeas* aug,
     std::string node(matches[i]);
     free(matches[i]);
 
-    Row r;
-    const char* value = nullptr;
-    int result = aug_get(aug, node.c_str(), &value);
-    if (result == 1) {
-      r["node"] = node;
-
-      if (value != nullptr) {
-        r["value"] = value;
-      }
-
-      if (!use_path) {
-        r["path"] = getSpanInfo(aug, node, context);
-      } else {
-        r["path"] = pattern;
-      }
-
-      r["label"] = getLabelInfo(aug, node, context);
-
-      results.push_back(r);
-    } else if (result < 1) {
+    const char *value = nullptr, *label = nullptr;
+    char* file = nullptr;
+    result = aug_ns_attr(aug, "matches", i, &value, &label, &file);
+    if (result == -1) {
       reportAugeasError(aug);
+      return;
+    }
+
+    std::string path;
+    if (file != nullptr) {
+      path = file;
+      path = path.substr(6);
+      // The caller is responsible for the matching memory.
+      free(file);
+    } else {
+      // The iterator is currently pointing to a folder so we extract the path from the node.
+      path = node.substr(6);
     }
+
+    Row r = {{"node", node},
+             {"value", value == nullptr ? "" : value},
+             {"path", path},
+             {"label", label == nullptr ? "" : label}};
+    results[i] = r;
   }
 
   // aug_match() allocates the matches array and expects the caller to free it.
   free(matches);
 }
 
-QueryData genAugeas(QueryContext& context) {
-  augeas* aug = aug_init(
-      nullptr, FLAGS_augeas_lenses.c_str(), AUG_NO_ERR_CLOSE | AUG_ENABLE_SPAN);
+class AugeasHandle {
+public:
+  augeas* aug;
+  bool error;
+
+  AugeasHandle() {
+    error = false;
+    this->aug = aug_init(nullptr, FLAGS_augeas_lenses.c_str(), AUG_NO_ERR_CLOSE | AUG_NO_LOAD);
+    // Handle initialization errors.
+    if (this->aug == nullptr) {
+      LOG(ERROR) << "An error has occurred while trying to initialize augeas";
+      error = true;
+    } else if (aug_error(this->aug) != AUG_NOERROR) {
+      error = true;
+      // Do not use aug_error_details() here since augeas is not fully
+      // initialized.
+      LOG(ERROR) << "An error has occurred while trying to initialize augeas: "
+              << aug_error_message(this->aug);
+      aug_close(this->aug);
+    }
+  }
 
-  // Handle initialization errors.
-  if (aug == nullptr) {
-    VLOG(1) << "An error has occurred while trying to initialize augeas";
-    return {};
-  } else if (aug_error(aug) != AUG_NOERROR) {
-    // Do not use aug_error_details() here since augeas is not fully
-    // initialized.
-    VLOG(1) << "An error has occurred while trying to initialize augeas: "
-            << aug_error_message(aug);
-    aug_close(aug);
+  ~AugeasHandle() {
+    aug_close(this->aug);
```

Comment:
Does `aug_close` check for a `nullptr`? If it doesn't then that check should happen here in the destructor.

Reply:
Alternatively, you could use a `std::shared_ptr` since the pointer is accessed outside of the class too. In the constructor you can pass a destructor lambda for the pointer.

Reply:
That's true.

Reply:
`aug_close` is perfectly happy with a `nullptr` ...